src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,906 @@
+     *
+     * @return NTLM Authenticate message
+     * @throws SQLServerException
+     */
+    private byte[] generateNtlmAuthenticate() throws SQLServerException {
+        int domainNameLen = null != context.domainBytes ? context.domainBytes.length : 0;
+        int userNameLen = null != context.userNameBytes ? context.userNameBytes.length : 0;
+        int workstationLen = null != context.workstationBytes ? context.workstationBytes.length : 0;
+        byte[] msg = null;
+
+        try {
+            // get NT challenge response
+            byte[] ntChallengeResp = getNtChallengeResp();
+            int ntChallengeLen = ntChallengeResp.length;
+
+            context.token = ByteBuffer
+                    .allocate(NTLM_AUTHENTICATE_PAYLOAD_OFFSET + NTLM_LMCHALLENAGERESPONSE.length
+                            + ntChallengeResp.length + domainNameLen + userNameLen + workstationLen)
+                    .order(ByteOrder.LITTLE_ENDIAN);
+
+            // set NTLM signature and message type
+            context.token.put(NTLM_HEADER_SIGNATURE, 0, NTLM_HEADER_SIGNATURE.length);
+            context.token.putInt(NTLM_MESSAGE_TYPE_AUTHENTICATE);
+
+            // start of payload data
+            int offset = NTLM_AUTHENTICATE_PAYLOAD_OFFSET;
+
+            // LM challenge response
+            context.token.putShort((short) 0);
+            context.token.putShort((short) 0);
+            context.token.putInt(offset);
+            offset += NTLM_LMCHALLENAGERESPONSE.length;
+
+            // NT challenge response
+            context.token.putShort((short) ntChallengeLen);
+            context.token.putShort((short) ntChallengeLen);
+            context.token.putInt(offset);
+            offset += ntChallengeLen;
+
+            // domain name fields
+            context.token.putShort((short) domainNameLen);
+            context.token.putShort((short) domainNameLen);
+            context.token.putInt(offset);
+            offset += domainNameLen;
+
+            // user name fields
+            context.token.putShort((short) userNameLen);
+            context.token.putShort((short) userNameLen);
+            context.token.putInt(offset);
+            offset += userNameLen;
+
+            // workstation fields
+            context.token.putShort((short) workstationLen);
+            context.token.putShort((short) workstationLen);
+            context.token.putInt(offset);
+            offset += workstationLen;
+
+            // not used = encrypted random session key fields
+            context.token.putShort((short) 0);
+            context.token.putShort((short) 0);
+            context.token.putInt(offset);
+
+            // same negotiate flags sent in negotiate message
+            context.token.putInt((int) context.negotiateFlags);
+
+            // version - not used but need to send blank to separate from MIC otherwise server confuses this with MIC
+            context.token.put(NTLMSSP_VERSION, 0, NTLMSSP_VERSION.length);

```

Comment:
This line is not required. The array is already initialized to zeroes, you just need to increment the index by 8. 

`context.token.position(context.token.position() + 8);`

Reply:
you're right, but is that any faster than putting 0's? I thought it makes it more clear..and in case we ever want to send version